require at least server version 10.8.0 to use HLS

mueslimak3r opened the pull request:
**Changes**
* added a method to return whether the current server meets a given minimum version requirement
  _it should probably be moved somewhere more suitable_
* exoplayer profile only adds its transcoding profile if HLS is enabled

* fixed transcoding profile still using `ContainsCodec(CodecTypes.HEVC, ContainerTypes.MP4)` instead of `ContainsCodec(CodecTypes.HEVC, ContainerTypes.TS)`
* removed unused variable in playback controller

**Issues**
* HLS transcoding is partially broken prior to 10.8.0 PR:
  https://github.com/jellyfin/jellyfin/pull/7338
  prior to that PR, HLS transcode streams may repeat the last few seconds of a video or not seek accurately
  
  **Notes**
Please suggest a better place to store the `getServerMeetsMinimumVersion()` method

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/playback/CustomPlaybackOverlayFragment.java:
```diff
@@ -1487,4 +1494,29 @@ private void renderSubtitles(@Nullable final String text) {
         });
     }
+
+    public boolean getServerMeetsMinimumVersion(@NonNull String minServerVersion) {
```
```suggestion
    public boolean getServerMeetsMinimumVersion(@NonNull ServerVersion minServerVersion) {
```
Use an instance of server version instead of a string, the constructor is easy to use. Should look something like `new ServerVersion(10, 8, 0);` from Java.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/playback/CustomPlaybackOverlayFragment.java:
```diff
@@ -1487,4 +1494,29 @@ private void renderSubtitles(@Nullable final String text) {
     }
+
+    public boolean getServerMeetsMinimumVersion(@NonNull String minServerVersion) {
+        if (sessionRepository.getValue().getCurrentSession().getValue() == null) {
```
Assign the session to a variable, the value might have changed after this check. (although unlikely)

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/playback/CustomPlaybackOverlayFragment.java:
```diff
@@ -1487,4 +1494,29 @@ private void renderSubtitles(@Nullable final String text) {
+            return false;
+        }
+
+        UUID serverId = sessionRepository.getValue().getCurrentSession().getValue().getServerId();
+        AuthenticationStoreServer server = authenticationStore.getValue().getServer(serverId);
+        if (server == null || server.getVersion() == null) {
+            Timber.d("server is null or version is null");
+            return false;
+        }
+
+        ServerVersion version = ServerVersion.Companion.fromString(server.getVersion());
+        ServerVersion minVersion = ServerVersion.Companion.fromString(minServerVersion);
+        if (version == null || minVersion == null) {
+            Timber.d("failed to create ServerVersions from strings: %s %s", version, minVersion);
+            return false;
+        }
+
+        int comparison = version.compareTo(minVersion);
+        Timber.d("current server version [%s] is %s [%s]", version, comparison >= 0 ? ">=" : "<", minVersion);
+        return comparison >= 0;
```
In AuthenticationRepository, replace the current getServers function with this:

```kotlin
	private fun AuthenticationStoreServer.asServer(id: UUID) = Server(
		id = id,
		name = name,
		address = address,
		version = version,
		loginDisclaimer = loginDisclaimer,
		dateLastAccessed = Date(lastUsed),
	)

	override fun getServers() = authenticationStore.getServers()
		.map { (id, entry) -> entry.asServer(id) }
		.sortedWith(serverComparator)

	override fun getServer(id: UUID) = authenticationStore.getServer(id)?.asServer(id)
```

(also add `fun getServer(id: UUID): Server?` to the interface, after getServers)

Then, in the server data class, change serverVersion to public or add a function to compare with a ServerVersion instance.

Then get the server via the AuthenticationRepository, or maybe even add a getServer() to the ServerRepository (although that feels a bit redundant..). And use that for the comparison here

nielsvanvelzen replied:
```patch
diff --git a/app/src/main/java/org/jellyfin/androidtv/auth/AuthenticationRepository.kt b/app/src/main/java/org/jellyfin/androidtv/auth/AuthenticationRepository.kt
index 61cfef7b3..cdfb9556b 100644
--- a/app/src/main/java/org/jellyfin/androidtv/auth/AuthenticationRepository.kt
+++ b/app/src/main/java/org/jellyfin/androidtv/auth/AuthenticationRepository.kt
@@ -31,6 +31,7 @@ import java.util.UUID
 
 interface AuthenticationRepository {
 	fun getServers(): List<Server>
+	fun getServer(id: UUID): Server?
 	fun getUsers(server: UUID): List<PrivateUser>?
 	fun getUser(server: UUID, user: UUID): PrivateUser?
 	fun saveServer(id: UUID, name: String, address: String, version: String?, loginDisclaimer: String?)
@@ -54,16 +55,20 @@ class AuthenticationRepositoryImpl(
 	private val serverComparator = compareByDescending<Server> { it.dateLastAccessed }.thenBy { it.name }
 	private val userComparator = compareByDescending<PrivateUser> { it.lastUsed }.thenBy { it.name }
 
-	override fun getServers() = authenticationStore.getServers().map { (id, info) ->
-		Server(
-			id = id,
-			name = info.name,
-			address = info.address,
-			version = info.version,
-			loginDisclaimer = info.loginDisclaimer,
-			dateLastAccessed = Date(info.lastUsed),
-		)
-	}.sortedWith(serverComparator)
+	private fun AuthenticationStoreServer.asServer(id: UUID) = Server(
+		id = id,
+		name = name,
+		address = address,
+		version = version,
+		loginDisclaimer = loginDisclaimer,
+		dateLastAccessed = Date(lastUsed),
+	)
+
+	override fun getServers() = authenticationStore.getServers()
+		.map { (id, entry) -> entry.asServer(id) }
+		.sortedWith(serverComparator)
+
+	override fun getServer(id: UUID) = authenticationStore.getServer(id)?.asServer(id)
 
 	private fun mapUser(
 		serverId: UUID,
```

mueslimak3r commented:
I added the changes suggested above.

I do still think it would be nice to move `getServerMeetsMinimumVersion` (now renamed) somewhere more appropriate

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/auth/model/Server.kt:
```diff
@@ -16,9 +16,16 @@ data class Server(
 	var dateLastAccessed: Date = Date(0),
 ) {
-	private val serverVersion = version?.let(ServerVersion::fromString)
+	val serverVersion = version?.let(ServerVersion::fromString)
```
Variable is not used outside of this class so it can be kept private

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/auth/model/Server.kt:
```diff
@@ -16,9 +16,16 @@ data class Server(
 	val versionSupported = serverVersion != null && serverVersion >= ServerRepository.minimumServerVersion
 
+
+	fun isVersionEqualOrGreater(otherVersion: ServerVersion?): Boolean {
```
I think we can make this an operator function [^1]. That way the Kotlin usages look a bit nicer.

[^1]: https://kotlinlang.org/docs/operator-overloading.html#comparison-operators

mueslimak3r replied:
like this?

Wouldn't a comparison overload be for comparing two objects of the same type?

`operator fun compareTo(other: ServerVersion?): Int  = serverVersion?.let { other?.compareTo(it) } ?: -1`

nielsvanvelzen replied:
The compareTo function is not restricted to the same type, same for other operators actually.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/auth/model/Server.kt:
```diff
@@ -16,9 +16,16 @@ data class Server(
 
+
+	fun isVersionEqualOrGreater(otherVersion: ServerVersion?): Boolean {
+		val serverVersion: ServerVersion? = version?.let { ServerVersion.fromString(it) }
```
Re-use serverVersion from the class, no need to parse it again

nielsvanvelzen replied:
(so just remove this line)

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/auth/model/Server.kt:
```diff
@@ -16,9 +16,16 @@ data class Server(
 	val versionSupported = serverVersion != null && serverVersion >= ServerRepository.minimumServerVersion
 
+
+	fun isVersionEqualOrGreater(otherVersion: ServerVersion?): Boolean {
```
Should be not-null

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/playback/CustomPlaybackOverlayFragment.java:
```diff
@@ -1487,4 +1494,25 @@ private void renderSubtitles(@Nullable final String text) {
+        boolean result = server.isVersionEqualOrGreater(minServerVersion);
+        Timber.d("current server version [%s] is %s [%s.%s.%s]",
+                server.getVersion(), result ? ">=" : "<",
+                minServerVersion.getMajor(), minServerVersion.getMinor(), minServerVersion.getPatch());
```
You should be able to just do `minServerVersion.toString()` to get the version as "10.8.0"

nielsvanvelzen commented:
I'm not the greatest fan of having getServerVersionEqualOrGreater in the CustomPlaybackOverlayFragment either. The next release (0.14) will require the server to be 10.8 or newer though, so we can remove it after 0.13 is released.
